### PR TITLE
feat: add chat history

### PR DESCRIPTION
**Pull request by ccheng26**
## Ticket

https://navalabs.atlassian.net/browse/DST-546

## Changes
- updated chainlit to v 1.1.306 to use `chat_context`
- revised `on_message` function to have `chat_history` param
- passed chat history to `generate` function

## Testing
- ask chatbot about previous query, chatbot should respond with additional information based on chat history
chathistory

### CODE REVIEW

File: app/src/generate.py [hunk 1]
```diff
@@ -72,11 +73,17 @@ def generate(
             },
         )
 
-    messages.append({"content": query, "role": "user"})
+    # chat_history has the user query as the last item, but we want to insert the context first
+    if chat_history:
+        chat_history.pop()
+        chat_history += messages
+    else:
+        chat_history = messages
```
Comment (yoomlam) [inline app/src/generate.py:81]: Could this be simplified to:
```suggestion
    if chat_history:
        chat_history.pop()
        messages.extend(chat_history)
```
and revert line 86?

Does the `chat_history` include repeated occurrences of the `system_prompt`?
Adding a unit test would help answer that.

Comment (ccheng26) [reply app/src/generate.py:81]: chat history doesn't include the `system_prompt` I'll add a test for that.

File: app/tests/src/test_generate.py [hunk 2]
```diff
@@ -112,6 +112,34 @@ def test_generate(monkeypatch):
...
+        {"role": "assistant", "content": "<div> answer</div>"},
+        {"content": "some other query", "role": "user"},
+    ]
+    expected_response = (
+        'Called gpt-4o with [{"content": "'
+        + PROMPT
+        + '", "role": "system"}, {"content": "Use the following context to answer'
+        + ' the question: context", "role": "system"},'
+        + ' {"content": "some query", "role": "user"}, {"content": "<div> answer</div>",'
+        + ' "role": "assistant"}, {"content": "some other query", "role": "user"}]'
```
Comment (yoomlam) [inline app/tests/src/test_generate.py:128]: Make this easier to read:
```suggestion
        'Called gpt-4o with [{"content": "' + PROMPT + '", "role": "system"}, "
        '{"content": "Use the following context to answer the question: context", "role": "system"}, '
        '{"content": "some query", "role": "user"}, '
        '{"content": "<div> answer</div>", "role": "assistant"}, '
        '{"content": "some other query", "role": "user"}]'
```

File: app/tests/src/test_generate.py [hunk 3]
```diff
@@ -124,3 +152,9 @@ def test_generate_with_context_with_score(monkeypatch, chunks_with_scores):
         + '", "role": "system"}, {"content": "some query", "role": "user"}]'
     )
     assert generate("gpt-4o", PROMPT, "some query", context_text) == expected_response
+
+
+[
+    {"role": "assistant", "content": "Michigan Bridges Eligibility Manual Chat Engine started "},
+    {"role": "user", "content": "some query"},
+]
```
Comment (yoomlam) [inline app/tests/src/test_generate.py:160]: stray code

File: app/tests/src/test_generate.py [hunk 4]
```diff
@@ -112,6 +112,34 @@ def test_generate(monkeypatch):
     assert generate("gpt-4o", PROMPT, "some query") == expected_response
 
 
+def test_generate_with_history(monkeypatch):
+    monkeypatch.setattr("src.generate.completion", mock_completion.mock_completion)
+    history = [
+        {"content": "some query", "role": "user"},
+        {"role": "assistant", "content": "<div> answer</div>"},
```
Comment (yoomlam) [inline app/tests/src/test_generate.py:119]: Is there a reason the keys are swapped?
```suggestion
        {"content": "<div> answer</div>", "role": "assistant"},
```

**Review by yoomlam — APPROVED** [review]

File: app/tests/src/test_generate.py (hunk 2 above)
Comment (ccheng26) [reply app/tests/src/test_generate.py]: this auto formats to 
```
'Called gpt-4o with [{"content": "'
        + PROMPT
        + '", "role": "system"}, "'
        + '{"content": "Use the following context to answer the question: context", "role": "system"}, '
        '{"content": "some query", "role": "user"}, '
        '{"content": "<div> answer</div>", "role": "assistant"}, '
        '{"content": "some other query", "role": "user"}]'
```

Comment (yoomlam) [reply app/tests/src/test_generate.py]: You can use an f-string for the first line to remove the `+`s.

**Pull request merged**